File: lib/models/proxy_config.dart
import 'package:equatable/equatable.dart';

/// LLM provider supported by jxproxy.
///
/// Aligned with the Swift [ProxyConfig.Provider] enum so routing config
/// is consistent across macOS and Android.
enum ProxyProvider {
  /// Anthropic Direct — own API key.
  direct,

  /// OpenRouter — routes through openrouter.ai.
  openrouter,

  /// OpenCode Zen / Big-Pickle.
  opencodeZen,

  /// OpenCode Go.
  opencodeGo,

  /// Google / Gemini.
  google,

  /// Nvidia NIM.
  nvidia,

  /// Nemotron 3 Ultra (via Nvidia).
  nemotron,

  /// Local / Ollama.
  local,

  /// Generic custom provider.
  custom;

  /// Human-readable label for UI dropdowns.
  String get displayName {
    switch (this) {
      case ProxyProvider.direct:
        return 'Anthropic Direct';
      case ProxyProvider.openrouter:
        return 'OpenRouter';
      case ProxyProvider.opencodeZen:
        return 'OpenCode Zen / Big-Pickle';
      case ProxyProvider.opencodeGo:
        return 'OpenCode Go';
      case ProxyProvider.google:
        return 'Google / Gemini';
      case ProxyProvider.nvidia:
        return 'Nvidia NIM';
      case ProxyProvider.nemotron:
        return 'Nemotron 3 Ultra';
      case ProxyProvider.local:
        return 'Local / Ollama';
      case ProxyProvider.custom:
        return 'Custom';
    }
  }
}

/// Log level for the jxproxy server.
enum ProxyLogLevel { debug, info, warn, error }

/// Configuration for the jxproxy router.
///
/// Mirrors the Swift [ProxyConfig] from the macOS app to keep shared
/// proxy settings portable across platforms.
class ProxyConfig extends Equatable {
  /// Port jxproxy listens on. Default: 5255.
  final int port;

  /// Primary LLM provider.
  final ProxyProvider provider;

  /// Model name/identifier to use with the provider.
  final String model;

  /// API key for the selected provider.
  final String apiKey;

  /// Custom host override (used by [ProxyProvider.local] and [ProxyProvider.custom]).
  final String customHost;

  /// Ordered list of fallback providers to try when the primary fails.
  final List<ProxyProvider> fallbackProviders;

  /// Log verbosity for jxproxy.
  final ProxyLogLevel logLevel;

  /// HTTP request timeout in seconds.
  final int requestTimeoutSeconds;

  const ProxyConfig({
    this.port = 5255,
    this.provider = ProxyProvider.opencodeZen,
    this.model = 'claude-sonnet-4-6',
    this.apiKey = '',
    this.customHost = '',
    this.fallbackProviders = const [],
    this.logLevel = ProxyLogLevel.info,
    this.requestTimeoutSeconds = 60,
  });

  ProxyConfig copyWith({
    int? port,
    ProxyProvider? provider,
    String? model,
    String? apiKey,
    String? customHost,
    List<ProxyProvider>? fallbackProviders,
    ProxyLogLevel? logLevel,
    int? requestTimeoutSeconds,
  }) {
    return ProxyConfig(
      port: port ?? this.port,
      provider: provider ?? this.provider,
      model: model ?? this.model,
      apiKey: apiKey ?? this.apiKey,
      customHost: customHost ?? this.customHost,
      fallbackProviders: fallbackProviders ?? this.fallbackProviders,
      logLevel: logLevel ?? this.logLevel,
      requestTimeoutSeconds: requestTimeoutSeconds ?? this.requestTimeoutSeconds,
    );
  }

  /// Whether the selected provider requires an API key.
  bool get requiresKey => provider != ProxyProvider.local;

  @override
  List<Object?> get props => [
    port, provider, model, apiKey, customHost,
    fallbackProviders, logLevel, requestTimeoutSeconds,
  ];
}

File: JXCODE/Models/ProxyConfig.swift
import Foundation

// MARK: - Proxy Configuration
// Merged from JXRouter's rich model with jxcode's minute provider configuration.

public struct ProxyConfig: Codable, Equatable {
    public var port: Int
    public var provider: Provider
    public var model: String
    public var fallbackProviders: [Provider]
    public var apiKeys: [Provider: String]
    public var logLevel: LogLevel
    public var autoStart: Bool
    public var launchAtLogin: Bool
    public var proxyAllTraffic: Bool
    public var requestTimeoutSeconds: Int

    /// System proxy settings
    public var httpProxy: String = ""
    public var httpsProxy: String = ""
    public var noProxy: String = ""
    public var allProxy: String = ""
    public var isProxyEnabled: Bool = false

    /// Custom host for local providers
    public var customHostUrl: String = ""

    public init(port: Int, provider: Provider, model: String, fallbackProviders: [Provider], apiKeys: [Provider: String], logLevel: LogLevel, autoStart: Bool, launchAtLogin: Bool, proxyAllTraffic: Bool, requestTimeoutSeconds: Int) {
        self.port = port
        self.provider = provider
        self.model = model
        self.fallbackProviders = fallbackProviders
        self.apiKeys = apiKeys
        self.logLevel = logLevel
        self.autoStart = autoStart
        self.launchAtLogin = launchAtLogin
        self.proxyAllTraffic = proxyAllTraffic
        self.requestTimeoutSeconds = requestTimeoutSeconds
    }

    public static let `default` = ProxyConfig(
        port: 5255,
        provider: .opencodeZen,
        model: "opencode/zen-coder",
        fallbackProviders: [.openrouter],
        apiKeys: [:],
        logLevel: .info,
        autoStart: false,
        launchAtLogin: false,
        proxyAllTraffic: false,
        requestTimeoutSeconds: 30
    )

    public enum Provider: String, Codable, CaseIterable, Identifiable {
        case direct = "Anthropic Direct"
        case openrouter = "OpenRouter"
        case opencodeZen = "OpenCode Zen / Big-Pickle"
        case opencodeGo = "OpenCode Go"
        case openai = "Nvidia NIM"
        case nemotron = "Nemotron 3 Ultra"
        case local = "Ollama Qwen"

        public var id: String { rawValue }

        public var identifier: String {
            switch self {
            case .direct: return "direct"
            case .openrouter: return "openrouter"
            case .opencodeZen: return "opencode-zen"
            case .opencodeGo: return "opencode-go"
            case .openai: return "openai"
            case .nemotron: return "openai"
            case .local: return "local"
            }
        }

        public var requiresKey: Bool { self != .local }

        public var defaultEndpoint: String {
            switch self {
            case .direct: return "https://api.anthropic.com"
            case .openrouter: return "https://openrouter.ai/api/v1"
            case .opencodeZen: return "https://opencode.ai/zen/v1"
            case .opencodeGo: return "https://opencode.ai/zen/go/v1"
            case .openai, .nemotron: return "https://integrate.api.nvidia.com/v1"
            case .local: return "http://127.0.0.1:11434"
            }
        }

        public var defaultModel: String {
            switch self {
            case .direct: return "claude-sonnet-5-20251001"
            case .openrouter: return "openrouter/auto"
            case .opencodeZen: return "opencode/zen-coder"
            case .opencodeGo: return "opencode/go-coder"
            case .openai: return "nvidia/llama-3.1-nemotron-70b-instruct"
            case .nemotron: return "nvidia/nemotron-3-8b-ultra"
            case .local: return "qwen2.5-coder:latest"
            }
        }
    }

    public enum LogLevel: String, Codable, CaseIterable, Identifiable {
        case debug, info, warn, error
        public var id: String { rawValue }
    }
}
